### DISCUSSION TITLE
What are Ocelot configuration and routing limitations?

### DISCUSSION THREAD

**Discussion by SheruGaur** [ThreeMammals/Ocelot#1795] [Q&A] [answered]
Hi,

We are internally doing a POC on ocelot as API Gateway in our organisation. 

While we really like the product till now, we had few basic queries. we didnt find much in documentation or the issues and discussion. 

If someone from their experience can share some info on following it will be helpful.

1. How many maximum routes can be configured in Ocelot.Json? 
2. Any limitations for define maximum routes in configuration file? 
3. Any standard configuration for which this has been tested?
4. in case of multiple configuration files, does Ocelot allow the change in individual config files to be updated in the merged ocelot.json at runtime?
5. Do we need to create an ocelot.global.json file for where multiple config files configuration to work?

Pls help with any references...

Thanks
Sheru

**Comment by raman-m** [comment]
@ks1990cn Tanmay, please help!
I believe you will explain to Mr. Sheru Ocelot [Configuration](https://ocelot.readthedocs.io/en/latest/features/configuration.html) feature, and answer on these questions better than me.

**Reply by SheruGaur** [reply]
@raman-m @ks1990cn 
Thanks for response, please provide solution for below mentioned points.

1. In case of multiple configuration files, does Ocelot allow the change in individual config files to be updated in the merged ocelot. Json at runtime?
2. Do we need to create an ocelot.global.json file for where multiple config files configuration to work?

**Reply by ks1990cn** [reply]
1.) Not sure what you are asking about, multiple files is just separate out your configurations routes (If you have large number of). Hope there will no duplicate routes defined in both files. You can think `.AddJsonFile` as loader and when you add `.AddOcelot`, it parse file to create configuration.
I will not change your changes if you have duplicates, you will get ambiguity with routes.

2.) As mentioned in Doc, **The GlobalConfiguration is a bit hacky and allows overrides of Route specific settings. It’s useful if you do not want to manage lots of Route specific settings**. No its not mandatory to have gloabal configuration.

**Reply by SheruGaur** [reply]
Hi @ks1990cn,

We have tried this approach before, here are the observations.

1. We have created two configuration files ocelot.common.json and ocelot.prd1.json 

2. The program.cs contains following code snippet:
```
static async Task Main(string[] args)
{
    await Host.CreateDefaultBuilder(args)
        .ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder.ConfigureAppConfiguration((hostingContext, config) =>
            {
                config
                    .SetBasePath(hostingContext.HostingEnvironment.ContentRootPath)
                    .AddJsonFile("appsettings.json", true, true)
                    .AddJsonFile($"appsettings.{hostingContext.HostingEnvironment.EnvironmentName}.json", true, true)
                    .AddOcelot(hostingContext.HostingEnvironment); // Using AddOcelot to merge Ocelot configuration files
            })
```
3. The generated ocelot.config is as below (not in a readable format :-():
```
{"Routes":[],"DynamicRoutes":[],"Aggregates":[],"GlobalConfiguration":{"RequestIdKey":null,"ServiceDiscoveryProvider":{"Scheme":null,"Host":null,"Port":0,"Type":null,"Token":null,"ConfigurationKey":null,"PollingInterval":0,"Namespace":null},"RateLimitOptions":{"ClientIdHeader":"ClientId","QuotaExceededMessage":null,"RateLimitCounterPrefix":"ocelot","DisableRateLimitHeaders":false,"HttpStatusCode":429},"QoSOptions":{"ExceptionsAllowedBeforeBreaking":0,"DurationOfBreak":0,"TimeoutValue":0},"BaseUrl":null,"LoadBalancerOptions":{"Type":null,"Key":null,"Expiry":0},"DownstreamScheme":null,"HttpHandlerOptions":{"AllowAutoRedirect":false,"UseCookieContainer":false,"UseTracing":false,"UseProxy":true,"MaxConnectionsPerServer":2147483647},"DownstreamHttpVersion":null}}
```
4. if we create a new file ocelot.global.json with only GlobalConfiguration and run then the ocelot.json is created properly as seen in attached files.
[ocelot.common.json](https://github.com/ThreeMammals/Ocelot/files/13448785/ocelot.common.json)
[ocelot.global.json](https://github.com/ThreeMammals/Ocelot/files/13448787/ocelot.global.json)
[ocelot.json](https://github.com/ThreeMammals/Ocelot/files/13448788/ocelot.json)
[ocelot.prd1.json](https://github.com/ThreeMammals/Ocelot/files/13448789/ocelot.prd1.json)

5. However we also see the files depicted as below
![image](https://github.com/ThreeMammals/Ocelot/assets/151524731/f1e1247a-c6d7-4761-8ec6-aa76d36b870b)

so if we try to delete ocelot.json before recompiling it deletes all other files also. so we have to rename ocelot.json with some other name and then delete it before retrying the build again.

6. And if want to update config on fly for any 1 product that also doesn't get reflected in ocelot.json which happens in single config file.

Please can you let us know if this is intended scenario or its a bug.

Thanks for your time.
Sheru

**Reply by ks1990cn** [reply]
@SheruGaur Give me time, I haven't read your queries... Need little time, busy these days.

**Reply by SheruGaur** [reply]
@ks1990cn Hi Tanmay, take your time but provide the prior solution for my raise queries.

**Reply by ks1990cn** [reply]
I am trying to replicate it, meanwhile look at https://stackoverflow.com/questions/67057050/cannot-merge-ocelot-config-files

**Reply by SheruGaur** [reply]
@ks1990cn  Still waiting for your reply.

**Reply by ks1990cn** [reply]
You tried this solution?

**Reply by SheruGaur** [reply]
Yes, but it is not working when any configuration change on the fly

**Comment by ks1990cn** [comment, marked as answer]
Hi @SheruGaur ,

Here are your answers:

1.) There is no maximum limit for routes you can define in Ocelot.Json. Internally code to uses simply list public List<Route> Routes { get; }. Just having huge number of routes with heavy configuration may impact startup/ search time for downstream route mapping. But I am pretty sure It can handle thousands / Lakhs without much impact on startup/map to downstream route.

2.) No there is no limit.

3.) There are few samples you can refer for what you are looking for, every where its tested ok. Ocelot have nice standards for test cases.
https://ocelot.readthedocs.io/en/latest/
https://learn.microsoft.com/en-us/dotnet/architecture/microservices/multi-container-microservice-net-applications/implement-api-gateways-with-ocelot
You can discuss with us, we can help you in understanding things you need.

**Reply by raman-m** [reply]
Thank you!

**Reply by SheruGaur** [reply]
thanks @ks1990cn. for providing the suggestions but still I am waiting for answer point no. 4th and 5th. Also, I want to know about the highest hardware configuration for ocelot Api gateway setup?

Thanks

**Reply by raman-m** [reply]
The hardware configuration can be the same as for regular ASP.NET Core 7+ web app. No limits!
It is better to think in terms of Requests per second... and scalability... Hardware requirements are vertical scalability!
It is better to focus on horizontal scalability!... On Kubernetes and a cloud...
So, you should think in terms of Kubernetes and regular Docker container hardware requirements.
